InterfacePython/TasmanianParticleSwarm.py

```diff
@@ -0,0 +1,264 @@
+        '''
+        iNumDims = self.getNumDimensions()
+        if (lfBoxLower.shape[0] != iNumDims):
+            raise InputError("lfBoxLower", "lfBoxLower.shape[0] should match the number of dimensions")
+        if (lfBoxUpper.shape[0] != iNumDims):
+            raise InputError("lfBoxUpper", "lfBoxUpper.shape[0] should match the number of dimensions")
+        pLibDTSG.tsgParticleSwarmState_InitializeParticlesInsideBox(self.pStatePntr, lfBoxLower, lfBoxUpper, c_char_p(random01.sType),
+                                                                    c_int(random01.iSeed), type_dream_random(random01.pCallable))
+
+def ParticleSwarm(pObjectiveFunction, iNumIterations, pInside, oParticleSwarmState, fInertiaWeight, fCognitiveCoeff,
+                  fSocialCoeff, random01 = DREAM.RandomGenerator(sType = "default")):
+    '''
+    Wrapper around TasOptimization::ParticleSwarm().
+
+    Runs iNumIterations of the particle swarm algorithm on an input state cParticleSwarmState to minimize the function
+    pObjectiveFunction over a domain specified by pInside. The parameters fInertiaWeight, fCognitiveCoeff, and fSocialCoeff
+    control the evolution of the algorithm. The parameter random01 controls the distribution of the particles.
+
+    pObjectiveFunction  : a Python lambda representing the objective function; it should take in two 2D NumPy arrays (x_batch, fval_batch)
+                          and produce no outputs; when it is called, it should write the result of applying the objective function
+                          to every row of x_batch to fval_batch; it is expected that x_batch.shape[1] = .getNumDimensions().
+    iNumIterations      : a positive integer representing the number iterations the algorithm is run.
+    pInside             : a Python lambda representing the function domain; it should take in one 2D NumPy array (x) and produce an
+                          integer (isInside); when called, it should return True if x is in the domain and False otherwise;
+                          it is expected that x.shape[0] = .getNumDimensions().
+    oParticleSwarmState : an instance of the Python ParticleSwarmState class; it will contain the results of applying the algorithm.
+    fInertiaWeight      : a double that controls the speed of the particles; should be in (0,1).
+    fCognitiveCoeff     : a double that controls how much a particle favors its own trajectory; usually in [1,3].
+    fSocialCoeff        : a double that controls how much a particle favors the swarm's trajectories; usually in [1,3].
+    random01            : a DREAM.RandomGenerator instance that produces floats in [0,1]
+
+    NOTE: np.frombuffer() is used below in order to avoid copying the data owned by the C pointers into the generated NumPy arrays.
+    '''
+    def cpp_obj_fn(num_dim, num_batch, x_batch_ptr, fval_ptr):
+        py_x_batch_ptr = cast(x_batch_ptr, POINTER(c_double * num_batch * num_dim))
```

Comment:
The calls to `cast()` `frombuffer()` and `resize()` can be combined into a single call to:
```
aX = np.ctypeslib.as_array(x_batch_ptr, (num_batch, num_dim))
```

Reply:
Changed here and in the other lambda in this function.